### ISSUE TITLE
Alacritty with Tmux half speed of Konsole with Tmux?

### ISSUE THREAD

**Issue by ryanerwin** [jwilm/alacritty#2587]
Which operating system does the issue occur on?

Arch Linux (rolling release), X11, KDE Plasma Desktop

Somewhat following up on:
https://github.com/jwilm/alacritty/issues/179
https://github.com/jwilm/alacritty/issues/125

Based on other threads, seems like `time find /usr` is the typical benchmark rather than `vtebench`... In any case, my results are consistent across almost all tests that I've run.

As you point out in the alatricity readme.md
> This means you won't find things like tabs or splits (which are best left to a window manager or terminal multiplexer)

I always use Tmux, and am finding that Tmux under Alacritty is consistently half the speed of Tmux under Konsole, even though running `find /usr` in Alacritty is faster than Konsole.

```
tmux on alacritty: 'find /usr' time: 3.234s, cpu: 72%
tmux on konsole: find /usr' time: 1.777s, cpu: 96%
```

What are things I could adjust in the Alacritty or Tmux configuration to get higher performance under when running under Alacritty?

**Comment by enizor** [comment]
I tested in on Ubuntu for Windows. (`ubuntu.exe` vs alacritty).
For some reason, `time find /usr` is ~8s on both, but when using `tmux` it drops to ~1.7s.
I checked the length of the results by piping into `wc -l`.

I noticed that piping also drops the time to 1.4s.
Maybe something in the like of 
> One reason why gnome-terminal is so fast is that I believe the VTE widget renders to a fixed frame rate in an almost vsync fashion and thus doesn't waste time rendering intermediate frames.

is happening here.

Tmux version: 2.6
Using my custom `.tmux.conf` or the default one did not change anything.

**Comment by chrisduerr** [comment]
> Based on other threads, seems like time find /usr is the typical benchmark rather than vtebench

Using `find` as a benchmark is terrible, since it is incredibly reliant on filesystem caches.

> I tested in on Ubuntu for Windows

The Windows filesystem is incredibly slow, making `find` an even worse benchmark there.

> One reason why gnome-terminal is so fast is that I believe the VTE widget renders to a fixed frame rate in an almost vsync fashion and thus doesn't waste time rendering intermediate frames.

So does Alacritty, not only 'in a vsync fashion' though, It uses vsync.

Please try reproducing this with a benchmark that does not rely on the filesystem so extensively. This is precisely what `vtebench` was written for, so please make use of it.

**Comment by ryanerwin** [comment]
> Please try reproducing this with a benchmark that does not rely on the filesystem so extensively. This is precisely what vtebench was written for, so please make use of it.

I thought it was odd that so many of the github issues were reference using `find /usr` instead of the `vtebench` tools, so I was just trying to be consistent with other posters.

```
'vtebench scrolling' time: 0.756s, cpu: 78% # alacritty under tmux
'vtebench scrolling' time: 0.379s, cpu: 99% # konsole under tmux

'vtebench scrolling-in-region' time: 0.670s, cpu: 86% # alacritty under tmux
'vtebench scrolling-in-region' time: 0.457s, cpu: 93% # konsole under tmux

# almost the same speed
'vtebench unicode-random-write' time: 3.213s, cpu: 99% # alacritty under tmux
'vtebench unicode-random-write' time: 3.290s, cpu: 99% # konsole under tmux

# alacritty was faster on this!
'vtebench alt-screen-random-write' time: 1.703s, cpu: 99% # alacritty under tmux
'vtebench alt-screen-random-write' time: 2.295s, cpu: 98% # konsole under tmux
```

For both terminals, the font is `Hack` patched with powerline symbols.

Of course, without Tmux, alacritty is considerably faster than Konsole.

**Comment by chrisduerr** [comment]
Do you use a laptop, if so, is it plugged in? Alacritty uses the GPU, so if your GPU is in low power mode that will likely have negative effects on Alacritty.

Also which GPU do you have? Is it an integrated one, or a dedicated GPU?

And can you reproduce these differences outside of tmux?

**Comment by ryanerwin** [comment]
> Do you use a laptop, if so, is it plugged in? 

Yes, [Dell XPS13 9350](https://wiki.archlinux.org/index.php/Dell_XPS_13_(9350))
Plugged in to power.

> Also which GPU do you have? Is it an integrated one, or a dedicated GPU?

Integrated Intel GPU
Intel i7-6560 CPU with Iris 540 graphics

> can you reproduce these differences outside of tmux?

No, outside of Tmux, alacritty is definitely faster.

```
'vtebench scrolling' time: 0.322s, cpu: 98% # alacritty no tmux
'vtebench scrolling' time: 3.667s, cpu: 16% # konsole no tmux

'vtebench scrolling-in-region' time: 0.358s, cpu: 98% # alacritty no tmux
'vtebench scrolling-in-region' time: 3.376s, cpu: 16% # konsole no tmux

'vtebench unicode-random-write' time: 3.272s, cpu: 99% # alacritty no tmux
'vtebench unicode-random-write' time: 18.463s, cpu: 29% # konsole no tmux

'vtebench alt-screen-random-write' time: 2.131s, cpu: 99% # alacritty no tmux
'vtebench alt-screen-random-write' time: 2.376s, cpu: 98% # konsole no tmux
```

I'm really pleasantly surprised at how fast alacritty is outside of Tmux!
But I use Tmux all the time for every window...
Sure hope there's a way to get the same speed while still using Tmux.

Are you using Tmux inside Alacritty and not seeing a slowdown compared to Konsole?

**Comment by chrisduerr** [comment]
Are you using the same configuration and terminfo for both Konsole and Tmux?

**Comment by ryanerwin** [comment]
Yes, both are linked to the same `/usr/share/terminfo` database.
For both, the `$TERM` is set to `xterm-256color`
For both, `$LANG` is set to `en_US.UTF-8`

<details>
  <summary>`locale output` on both alacritty and konsole identical</summary>

```
LANG=en_US.UTF-8
LC_CTYPE=zh_CN.UTF-8
LC_NUMERIC=en_US.UTF-8
LC_TIME=en_US.UTF-8
LC_COLLATE=en_US.UTF-8
LC_MONETARY=en_US.UTF-8
LC_MESSAGES=en_US.UTF-8
LC_PAPER=en_US.UTF-8
LC_NAME=en_US.UTF-8
LC_ADDRESS=en_US.UTF-8
LC_TELEPHONE=en_US.UTF-8
LC_MEASUREMENT=en_US.UTF-8
LC_IDENTIFICATION=en_US.UTF-8
LC_ALL=
```
</details>

Normally I use my laptop connected to an external display, on AC power.
Whether I used the large external display or small internal display, the ratio (vtebench under Tmux faster on Konsole than Alacritty).

Running tmux with `tmux -f/dev/null` did not change the results.

Seems so strange that even with my relatively weak GPU, Alacritty is much faster than Konsole until I run Tmux on both (my normal use case) and then it reverses...

Any thoughts on how to further troubleshoot?
Have any other users reported this? Maybe it's Arch specific?

**Comment by PowerSnail** [comment]
In my case, Alacritty outperformed both Gnome Terminal and Konsole outside Tmux, and at least equal to them inside Tmux (within margin of error).

![output](https://user-images.githubusercontent.com/7428673/60085553-5d1bda00-96ee-11e9-9975-81dc0d4a8b15.png)

I guess this would depend on the computing power of the specific CPU and GPU, wouldn't it? I use a RTX 2080; that probably helped a lot in this case.

**Comment by ryanerwin** [comment]
> computing power of the specific CPU and GPU, wouldn't it? 

You must be correct. Obviously my `Iris Graphics 540` doesn't compare with an RTX 2080...

That said, it's still very surprising that even with a weak GPU, `altracitty` outside of Tmux blows away Konsole, but inside of Tmux I get the opposite result.

Sounds like the answer is somewhere in the details of how Tmux actually works, but since Tmux is the recommended method of doing "tabs" in alacritty, would be fantastic to get to the bottom of it.

Please let me know if you've to any suggestions on where to look...

**Comment by chrisduerr** [comment]
@ryanerwin One idea I have is that your benchmarking times are extremely small, that's not really beneficial to testing actual performance since lots of other things take up a relatively large amount of time compared to longer benchmarks (file opening/output/etc).

Could you test this again with benchmarks that take around 10-15s? You can control the number of bytes vtebench uses with the `-b` option during generation. With the alt-screen-rando-write benchmark you'll also need to write to a file and then do `time cat file` since Alacritty is faster than vtebench outputs on that benchmark (without tmux at least).

I've tested this with urxvt and both it and Alacritty were slowed down from ~6s down to ~15s by tmux, making it seem like this is mostly up to tmux slowing things down and not the terminal emulator anymore. I'd chalk up the difference inside of tmux between Alacritty/urxvt up to margin of error, though Alacritty was a little bit faster. For reliable results this would have to be tested 10+ times with switching between emulators between tests.

**Comment by kchibisov** [comment]
I was a bit curios about this bottleneck, so I've decided to run some profiling here.  So if you take a look at records[1] and compare `perf.data.alacritty` and `perf.data.kitty` you'll find them mostly the same. These profiles are flooded with `tmux` stuff rather then `alacritty`s/`kitty`s ( this is something expected though).  So, `tmux` seems like a bottleneck in this benchmark (it also bumps one core to 100% on my machine during benchmark) and I think if we want to get performance benefits from our super high throughput in `tmux`, `tmux` should be improved somehow, rather then tuning `alacritty` to please `tmux` here.

I've also tested `screen` and profile was similar to `tmux` one, but screen was slower.

--
Benchmark for profiling
```
perf stat -r 10 vtebench -b 10000000 scrolling
```

[records](https://www.dropbox.com/sh/m9bpauosq40lyih/AACpzmtuWi5TC9OE_3zwH5rEa?dl=0) from `perf record alacritty`/`perf record kitty` [1]

**Comment by chrisduerr** [comment]
Since this is likely relevant information for everyone interested, I'd like to point out that these issues are likely because tmux's parser is too slow. Alacritty's parser actually performs exceptionally well but still is the main bottleneck for something like the asrw benchmark at the moment. So since tmux needs to implement a complete terminal emulator self (as a multiplexer it's pretty much a terminal emulator with a terminal UI), a slow parser will choke it right up regardless of rendering performance.

As far as I'm concerned there's nothing Alacritty can do about this. If you run multiple terminal emulators inside of each other, it just doesn't matter how fast the fastest one is. And they'll never be all the same speed, so there's really no way around hitting some performance bottleneck here.

**Comment by jbg** [comment]
Isn't that a pretty strong argument for implementing tabs in Alacritty so that there's no need to run multiple terminal emulators inside of each other?

**Comment by chrisduerr** [comment]
I wouldn't say so. Just for tabs, something like tmux is completely overkill. WMs are sufficiently capable of taking care of this, without taking any performance hit, so that's the place you should be looking for tabs imho.

**Comment by nicm** [comment]
Scrolling with tmux (after 2.4) in the presence of SD (indn in terminfo(5)) works very differently from without tmux because tmux will batch scrolling into sets of SD sequences rather than scrolling every individual line. Even without SD, it will batch them into sets of linefeeds. So running with and without tmux is probably not hitting the same code in the outer terminal.

tmux will also try to avoid sending data to the terminal that is later going to be scrolled off the screen, so it would be expected even on a fast terminal that tmux would be taking up a larger proportion of the CPU time, because it would be processing more of the data than the outer terminal ever sees. For example, running "seq 10000" inside tmux sends only about 500 lines to the outer terminal or about 10-15% of the data volume.

tmux will also check aggressively to make sure the terminal can keep up and switch to periodically redrawing the entire terminal if it can't.

This all doesn't help particularly on fast terminals and fast networks but it makes a big difference on slow ones (like xterm) - and since tmux has to process everything regardless to update its own internal state, there is no point in sending it even if the terminal can keep up. The focus so far has been on making slow terminals acceptable (particularly interactively) rather than making acceptable terminals faster.

I don't doubt the parser could be improved however - if anyone plans to look at this then feel free to get in touch.

There are other things tmux does that could probably also be improved by someone bent on optimization. For example, it tries to leave the terminal in a neutral state (no scroll region, SGR0) before blocking, so that a disconnect (eg ~. in ssh) doesn't leave the user upset - I don't know how much overhead this adds, but it could probably be reduced.

**Comment by nicm** [comment]
A 100% slowdown as in the numbers above when using tmux seems OK considering there are now two terminals (tmux plus alacritty) rather than one involved, so I don't think tmux is doing much harm here.

I see roughly the same: with a million lines of output on localhost, alacritty and rxvt are pretty comparable both with and without tmux and both seem acceptable (~1.5 seconds without tmux, ~2.5 seconds with tmux - xterm is ~21 seconds without and ~3 seconds with).

They are also both pretty fast with left-right split panes, which is good considering that they don't support margins, and tmux stays usable in one pane while the other is showing fast output.

So I don't really think there is much to fault tmux for here, although if anyone wants to work on performance of course that is welcome :-).

**Comment by kchibisov** [comment]
@nicm Just out of curiosity, is there a way that alacritty can help tmux to "update faster"? (not only draw tbh). I've seen your comment on #160 https://github.com/alacritty/alacritty/issues/160#issuecomment-573542489 , but maybe there's something more that we can do here?

**Comment by nicm** [comment]
Sorry for the delay in replying, I was on holiday last week.

I'm not sure there is anything else simple the terminal can do, I don't really have any ideas, apart from support for DECSLRM.

tmux will use some other stuff like DECFRA to erase but it can only be used for non-default background colours so it probably doesn't make any difference for most people.

DECSLRM makes a big difference with left-right splits simply because of the volume of data that tmux has to send, without DECSLRM tmux has to redraw the entire pane when it scrolls, which can mean hundreds of kilobytes of extra data.

**Comment by kchibisov** [comment]
Thanks for your reply, @nicm.

Well, we discussed adding DECSLRM on IRC and we like the idea behind it(we can add create better selection rotation, url tracking), however it's really hard to add it to alacritty, without losing performance (margin on a ring buffer isn't a thing tbh), so it'll likely take us some time to create a working solution (introducing performance regressions is not an option for alacritty).

As for this issue, I'll try to benchmark some our parts under tmux, maybe something isn't performing well, however from my testing, performance of most terminals with tmux is the same in numbers and just a matter whether outer terminal(e.g. alacritty) skipping frames or not(as you've already mentioned), so it's really hard to compare things here, but maybe there's our fault somewhere and  we can do curtain things even faster.

**Comment by nicm** [comment]
Sounds great, let me know if you add it.

**Comment by wiz21b** [comment]
To add a faint data point... In my own special benchmark (a TUI rendered 20fps with rust "tui" library), I have some interesting behavior. The CPU used by my program is 7% with alacritty, 10% with konsole (so somehow alacritty helps here, and by a significant margin; but the program is multithread => it could be some active waiting disappearing because of ideal conditions)). However the CPU used by alacritty is 46% whereas it's about 10% in konsole. So alacritty works a lot more than konsole. Maybe a gfx driver issue ?

According to alacritty info, the redraw time is +/- 230 usec, rather stable. The log says the glyph cache is initialized (which I guess means that my opengl hardware is runnning ok). I have a NVIDIA GPU GeForce GT 730 (GK208), running debian + kde.

Hope it helps...

**Comment by leet0rz** [comment]
@chrisduerr When are we getting tabs for linux and windows? Its only Mac that has tabs now right?

**Comment by kchibisov** [comment]
Tabs should be provided by window manager. macOS provides us with them, so we use them, windows/linux not, so no tabs.
